akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala

```diff
@@ -57,68 +57,82 @@ object WebSocketClientBlueprint {
-                  require(parseResult == ParserOutput.MessageEnd, s"parseResult should be MessageEnd but was $parseResult")
-                  parser.onPull() match {
-                    case NeedMoreData          ⇒ ctx.pull()
-                    case RemainingBytes(bytes) ⇒ ctx.push(bytes)
-                    case other ⇒
-                      throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
+            parser.setContextForNextResponse(HttpResponseParser.ResponseContext(HttpMethods.GET, None))
+
+            def onPush(): Unit = {
+              val elem = grab(in)
+              parser.parseBytes(elem) match {
+                case NeedMoreData ⇒ pull(in)
+                case ResponseStart(status, protocol, headers, entity, close) ⇒
+                  val response = HttpResponse(status, headers, protocol = protocol)
+                  Handshake.Client.validateResponse(response, subprotocol.toList, key) match {
+                    case Right(NegotiatedWebSocketSettings(protocol)) ⇒
+                      result.success(ValidUpgrade(response, protocol))
+
+                      become(transparent)
+                      valve.open()
+
+                      val parseResult = parser.onPull()
+                      require(parseResult == ParserOutput.MessageEnd, s"parseResult should be MessageEnd but was $parseResult")
+                      parser.onPull() match {
+                        case NeedMoreData          ⇒ pull(in)
+                        case RemainingBytes(bytes) ⇒ push(out, bytes)
+                        case other ⇒
+                          throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
+                      }
+                    case Left(problem) ⇒
+                      result.success(InvalidUpgradeResponse(response, s"WebSocket server at $uri returned $problem"))
+                      failStage(new IllegalArgumentException(s"WebSocket upgrade did not finish because of '$problem'"))
                   }
-                case Left(problem) ⇒
-                  result.success(InvalidUpgradeResponse(response, s"WebSocket server at $uri returned $problem"))
-                  ctx.fail(new IllegalArgumentException(s"WebSocket upgrade did not finish because of '$problem'"))
+                case other ⇒
+                  throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
               }
-            case other ⇒
-              throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
+            }
           }
-        }
-      }
 
-      def transparent: State = new State {
-        def onPush(elem: ByteString, ctx: Context[ByteString]): SyncDirective = ctx.push(elem)
-      }
+          def transparent: InHandler = new InHandler {
+            val elem = grab(in)
+            def onPush(): Unit = push(out, elem)
+          }
+          setHandlers(in, out, this)
```

Comment:
hm not sure if that's valid. this should mimic what `initial` was doing. Do we have the same logic invoked `initially`?